Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Maps.hs

```diff
@@ -43,9 +51,60 @@ autoComplete personId req = do
 getPlaceDetails :: (EncFlow m r, EsqDBFlow m r, CacheFlow m r, CoreMetrics m) => Id DP.Person -> Maps.GetPlaceDetailsReq -> m Maps.GetPlaceDetailsResp
 getPlaceDetails personId req = do
   person <- QP.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
+
   Maps.getPlaceDetails person.merchantId req
 
 getPlaceName :: (EncFlow m r, EsqDBFlow m r, CacheFlow m r, CoreMetrics m) => Id DP.Person -> Maps.GetPlaceNameReq -> m Maps.GetPlaceNameResp
 getPlaceName personId req = do
+  case req.getBy of
+    MIT.ByLatLong (Maps.LatLong lat lon) -> do
+      let myGeohash = DG.encode 8 (lat, lon)
+      let geoHash = pack (fromMaybe "" myGeohash)
+      placeNameCache <- QM.findPlaceByGeoHash geoHash
+      if null placeNameCache
+        then callMapsApi personId req
+        else mapM convertToGetPlaceNameResp placeNameCache
+    MIT.ByPlaceId placeId -> do
+      placeNameCache <- CM.findPlaceByPlaceId placeId
+      if null placeNameCache
+        then callMapsApi personId req
+        else mapM convertToGetPlaceNameResp placeNameCache
+
+callMapsApi :: (EncFlow m r, EsqDBFlow m r, CacheFlow m r, CoreMetrics m) => Id DP.Person -> Maps.GetPlaceNameReq -> m Maps.GetPlaceNameResp
+callMapsApi personId req = do
   person <- QP.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
-  Maps.getPlaceName person.merchantId req
+  res <- Maps.getPlaceName person.merchantId req
+  placeNameCache <- convertResultsRespToPlaceNameCache (head res)
+  let placeId = fromMaybe "" placeNameCache.placeId
+  let geoHash = fromMaybe "" placeNameCache.geoHash
+  CM.cachedPlaceByPlaceId placeId [placeNameCache]
+  CM.cachedPlaceByGeoHash geoHash [placeNameCache]
+  Esq.runTransaction $ QM.create placeNameCache
+  return res
```

Comment:
hm this is strange logic. Why do you cache only first result, but return all list?

Reply:
Actually at UI, we just use the first element of this response. So, even if we are giving whole response, UI is still ignoring the other elements.